crates/wasmcloud-provider-macros/src/lib.rs

```diff
@@ -0,0 +1,371 @@
+                    }
+                    "Message.RequestMulti" => {
+                        let input: RequestMultiBody = provider_sdk::deserialize(&body)?;
+                        let result = self
+                            .request_multi(
+                                ctx,
+                                input.subject,
+                                input.body,
+                                input.timeout_ms,
+                                input.max_results,
+                            )
+                            .await
+                            .map_err(|e| {
+                                provider_sdk::error::ProviderInvocationError::Provider(e.to_string())
+                            })?;
+                        Ok(provider_sdk::serialize(&result)?)
+                    }
+                    "Message.Publish" => {
+                        let input: PublishBody = provider_sdk::deserialize(&body)?;
+                        let result = self.publish(ctx, input.msg).await.map_err(|e| {
+                            provider_sdk::error::ProviderInvocationError::Provider(e.to_string())
+                        })?;
+                        Ok(provider_sdk::serialize(&result)?)
+                    }
+                    _ => Err(provider_sdk::error::InvocationError::Malformed(format!(
+                        "Invalid method name {method}",
+                    ))
+                             .into()),
+                }
+            }
+        }
+
+        // NOTE(thomastaylor312): In order for the provider to get what it needs to do work, it needs to
+        // pass the context with the actor ID on it (which means the signatures are slightly different than
+        // what is expressed in wit, but will be entirely contained to our generators). I'm guessing we can
+        // come up with a cleaner way to maybe do this so we aren't generating special functions here, but
+        // not going to do it for now.
+        #[async_trait]
+        pub trait Consumer {
+            async fn request(
+                &self,
+                ctx: wit_bindgen_wasmcloud::provider_sdk::Context,
+                subject: String,
+                body: Option<Vec<u8>>,
+                timeout_ms: u32,
+            ) -> Result<BrokerMessage, String>;
+
+            async fn request_multi(
+                &self,
+                ctx: wit_bindgen_wasmcloud::provider_sdk::Context,
+                subject: String,
+                body: Option<Vec<u8>>,
+                timeout_ms: u32,
+                max_results: u32,
+            ) -> Result<Vec<BrokerMessage>, String>;
+
+            async fn publish(&self, ctx: wit_bindgen_wasmcloud::provider_sdk::Context, msg: BrokerMessage)
+                             -> Result<(), String>;
+        }
+
+        /// Handle Messaging methods that interact with redis
+        #[async_trait]
+        impl Consumer for #impl_struct_name {
```

Comment:
Shouldn't need this either! This is what each provider implementation will implement

Reply:
Yep, this was not necessary -- removed